server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -1994,6 +1997,10 @@ private void validateRootVolumeDetachAttach(VolumeVO volume, UserVmVO vm) {
 
     @ActionEvent(eventType = EventTypes.EVENT_VOLUME_DETACH, eventDescription = "detaching volume")
     public Volume detachVolumeViaDestroyVM(long vmId, long volumeId) {
+        Account caller = CallContext.current().getCallingAccount();
+        Volume volume = _volsDao.findById(volumeId);
+        // Permissions check
+        _accountMgr.checkAccess(caller, null, true, volume);
```

Comment:
@lujiefsi VM destroy might fail if the caller cannot access any volume(s) attached to VM. So, I think it is better to check the access before destroying VM.

Reply:
destroyVolume alerady have an access check, in method "retrieveAndValidateVolume"

Reply:
detachVolumeFromVM still needs the access check to precent others detach a volume that does not belong to him

Reply:
@lujiefsi I mean, access check for all the volumes before destroying VM.

Reply:
@sureshanaparti  got it and see UserVmManagerImpl

Reply:
thanks @lujiefsi 
```suggestion
        _accountMgr.checkAccess(caller, null, true, volume);
```